home-manager/emacs/config.org
#+TITLE: Ohne's GNU/Emacs Configuration
#+AUTHOR: Ohne 
#+DESCRIPTION: Plain GNU/Emacs Configuration

* Package Management
MELPA (Package Archive) and default package.el management
#+begin_src emacs-lisp

(require 'package)
			
(add-to-list 'package-archives
	'("MELPA" . "https://melpa.org/packages/"))

(package-initialize)

(unless (package-installed-p 'use-package)
	(package-refresh-contents)
	(package-install 'use-package))

(use-package magit
	:ensure t)

#+end_src

* Appearance
TODO: Modeline, Scaling
** Themes
Doom themes
#+begin_src emacs-lisp

(use-package doom-themes
	:if window-system
	:ensure t
	:config
	(doom-themes-org-config)
	(doom-themes-visual-bell-config))

(load-theme 'doom-nord t)

#+end_src

** Font
Intel One Mono
#+begin_src emacs-lisp

(when (member "JetBrains Mono" (font-family-list))
	(set-frame-font "JetBrains Mono-18" t t))

(setq font-lock-maximum-decoration t)

#+end_src 

** UI
Cleanup
#+begin_src emacs-lisp

(setq use-dialog-box nil)
(setq use-file-dialog nil)
(setq make-backup-files nil)
(setq auto-save-default nil)
(setq ivy-use-selectable-prompt t)
(menu-bar-mode -1)
(tool-bar-mode -1)
(fringe-mode -1)
(scroll-bar-mode -1)
(global-subword-mode 1)

#+end_src

** Icons
All The Icons
#+begin_src emacs-lisp

(use-package all-the-icons
	:ensure t
	:if (display-graphic-p))

#+end_src

** Line Numbers
Relative line numbers
#+begin_src emacs-lisp

;; (when (version<= "26.0.50" emacs-version )
	;; (global-display-line-numbers-mode)
	;; (setq display-line-numbers-type 'relative))

#+end_src

** Modeline
Doom-modeline.el 
#+begin_src emacs-lisp

(use-package doom-modeline
	:init
	:ensure t
	:config
	(doom-modeline-mode 1))

(setq doom-modeline-bar-height 5)
(setq doom-modeline-bar-width 1)
(setq doom-modeline-buffer-file-name-style 'truncate-upto-root)
(setq doom-modeline-icon nil)
;; (setq doom-modeline-major-mode-icon t)
;; (setq doom-modeline-major-mode-color-icon t)
;; (setq doom-modeline-minor-modes nil)
(setq display-time-mode t)
(setq doom-modeline-time-icon t)
(setq doom-modeline-time t)
(setq doom-doom-modeline-buffer-name t)
(setq doom-modeline-lsp t)
(setq doom-modeline-buffer-encoding t)
(display-time)

#+end_src

** Cursor
Hl-column/line.el
#+begin_src emacs-lisp


(use-package hl-line
	:ensure t
	:config
	(global-hl-line-mode))

(use-package hl-column
	:ensure t
	:config)

#+end_src
Patch: Org-mode and dashboard breaks hl-column
#+begin_src emacs-lisp

(add-hook 'org-mode-hook (lambda () (hl-column-mode 0)))
(add-hook 'dashboard-mode-hook (lambda () (hl-column-mode 0)))

#+end_src 

** Dashboard
Dashboard.el
#+begin_src emacs-lisp

(use-package page-break-lines
  :ensure t)

(use-package dashboard
  :ensure t
  :config
  (dashboard-setup-startup-hook)
  (setq dashboard-banner-logo-title "in vino veritas")
  ;; (setq dashboard-startup-banner "~/.config/emacs/images/emacs.png")
  (setq dashboard-page-separator "\n\f\n"))

(global-set-key (kbd "M-o") 'dashboard-open)

#+end_src

** Size
Size and zoomzoom
#+begin_src emacs-lisp

(set-face-attribute 'default nil :height 170)

;; (defun increase-emacs-size () 
;;	(set-face-attribte 'default nil :height (+ x 20)))

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)

#+end_src

* QOL
Quality of Life
** File Completion
Ivy/Counsel.el
#+begin_src emacs-lisp

(electric-pair-mode 1)

(use-package ivy
	:init
	(ivy-mode)
	:ensure t
	:config
	(setq ivy-use-virtual-buffers t)
	(setq enable-recursive-minibuffers t))

(use-package counsel
	:init
	:ensure t)

#+end_src

** Indents/Tabs
Patch: tabs not working ootb
#+begin_src emacs-lisp

(delete-selection-mode 1)
(setq tab-width 4)
(setq-default tab-width 4)
(setq-default indent-tabs-mode nil)

(defun my-insert-tab-char ()
	(interactive)
		(insert "\t"))

(global-set-key (kbd "TAB") 'my-insert-tab-char)

#+end_src

** Binds
Patch: Enable C-u scrolling with Evil
#+begin_src emacs-lisp

(setq evil-want-C-u-scroll t)

#+end_src 
Evil.el 
#+begin_src emacs-lisp

(use-package evil
	:init
	(evil-mode)
	:ensure t
	:config)

#+end_src 
Custom Functions
#+begin_src emacs-lisp

(defun oh/insert-line-below ()
	(interactive)
	(save-excursion
	(end-of-line)
	(open-line 1)))
;
(defun oh/insert-line-above ()
	(interactive)
	(save-excursion
	(end-of-line 0)
	(open-line 1)))

(defun oh/insert-elisp-source-block ()
	(insert "#+begin_src emacs-lisp\n\n#+end_src"))

    #+end_src
General Global Binds
#+begin_src emacs-lisp

(global-set-key (kbd "C-;") 'comment-region)

#+end_src
General.el
#+begin_src emacs-lisp

(use-package general
	:ensure t
	:config
	(general-evil-setup)

	(general-create-definer oh/leader-keys
	:states '(normal insert visual emacs)
	:keymaps 'override
	:prefix "SPC" 
	:global-prefix "M-SPC")
	
  	(oh/leader-keys
    "." '(find-file :wk "Find File")
    "f c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Config")
    "f r" '(counsel-recentf :wk "Find recent files")
	"f s" '(save-buffer :wk "Save file"))

  	(oh/leader-keys
    "b" '(:ignore t :wk "Buffer")
    "b b" '(switch-to-buffer :wk "Switch buffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer")
	"b o" '(evil-buffer-new :wk "New buffer"))
	
	(oh/leader-keys
	"b m" '(bookmark-set :wk "Set bookmark")
	"b M" '(bookmark-delete :wk "Delete bookmark")
	"RET" '(bookmark-jump :wk "Jump to bookmark"))

 	(oh/leader-keys
    "h" '(:ignore t :wk "Help")
    "h f" '(describe-function :wk "Describe function")
    "h v" '(describe-variable :wk "Describe variable")
    "h r r" '(reload-init-file :wk "Reload emacs config")
	"h t" '(load-theme :wk "Load doom theme"))

	(oh/leader-keys
    "t" '(:ignore t :wk "Line Manipulation")
    "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
    "t t" '(visual-line-mode :wk "Toggle truncated lines")
    "t r" '((lambda () (setq display-line-numbers 'relative)) :wk "Toggle line numbers"))

	(oh/leader-keys
	"l" '(:ignore t :wk "Elisp")
	"l e" '(pp-eval-last-sexp :wk "Eval last sexp"))

)

#+end_src
Bind/Key Completion
#+begin_src emacs-lisp

(use-package which-key
	:init
	:ensure t
	:config
	(which-key-mode))

#+end_src 

* Development

** Org (org-mode)
Org.el
#+begin_src emacs-lisp

(use-package org
    :init
    :ensure t
    :config
    (add-hook 'org-mode-hook 'org-indent-mode))

(use-package org-modern
    :after org
    :ensure t
    :config
    (add-hook 'org-mode-hook #'org-modern-mode))

(use-package org-roam
  :ensure t
  :bind (("C-c n l" . org-roam-buffer-toggle)
         ("C-c n f" . org-roam-node-find)
         ("C-c n g" . org-roam-graph)
         ("C-c n i" . org-roam-node-insert)
         ("C-c n c" . org-roam-capture))
  :custom
  (org-roam-capture-templates
	'(
	  ("g" "generic" plain
	  "#+TITLE:  %?"
	  :if-new (file "~/Git/notes/${slug}-%<%m-%d-%Y>.org")
	  :unnarowed t)
         ("j" "journal" plain 
	  (file "~/.dotfiles/home-manager/emacs/templates/journal.org")
	  :if-new (file "~/Git/notes/journal-%<%m-%d-%Y>.org")
	  :unnarowed t)
         )
  :config
  (org-roam-db-autosync-mode)
  (require 'org-roam-protocol)))

(setq org-roam-directory "~/Git/notes") ;; errors when put inside use-package :custom module

;; (use-package org-roam-ui
    ;; :ensure t)

#+end_src
Patch: electric-indent-mode and org-edit-src-content-indentation breaking babel indentation
#+begin_src emacs-lisp

(add-hook 'org-mode-hook 
	(lambda () (electric-indent-local-mode -1)))

(setq org-edit-src-content-indentation 0)

#+end_src

** LSP
lsp-mode
#+begin_src emacs-lisp

(use-package lsp-mode
 	:ensure t)

(use-package lsp-ui
 	:ensure t
 	:commands lsp-ui-mode
 	:hook
 	(lsp-mode . lsp-ui-mode))

(defun org-babel-edit-prep:c (babel-info)
  (setq-local buffer-file-name (->> babel-info caddr (alist-get :tangle)))
  (lsp))

#+end_src
C/C++
#+begin_src emacs-lisp

(use-package ccls
  :ensure t
  :config
  (setq ccls-executable "ccls")
  (setq lsp-prefer-flymake nil)
  (setq-default flycheck-disabled-checkers '(c/c++-clang c/c++-cppcheck c/c++-gcc))
  :hook ((c-mode c++-mode objc-mode) .
         (lambda () (require 'ccls) (lsp))))

#+end_src
Python
#+begin_src emacs-lisp

(use-package lsp-pyright
	:ensure t
	:hook (python-mode . (lambda () (require 'lsp-pyright) (lsp))))

#+end_src

** Completion
Company.el
#+begin_src emacs-lisp

(use-package company
 	:ensure t)

;; (use-package company-lsp
;; 	:ensure t
;; 	:commands company-lsp
;; 	:config
;; 	(push 'company-lsp company-backends))

(use-package flycheck
	:ensure t)

(use-package capf-autosuggest
	:ensure t)

#+end_src

** Project Management
Projectile.el
#+begin_src emacs-lisp

;; Remove parenthesis when done
(use-package projectile
	:ensure t)
	;; :config
	;; (setq projectile-project-search-path '("~/Projects/" )))

#+end_src

** Sudo
Sudo-edit.el
#+begin_src emacs-lisp

;; (use-package sudo-edit
;; 	:config
;; 	(oh/leader-keys
;;   	"fu" '(sudo-edit-find-file :wk "Sudo find file")
;;   	"fU" '(sudo-edit :wk "Sudo edit file")))

#+end_src 

* Misc
Miscellaneous
** Time
Timezone
#+begin_src emacs-lisp

(set-time-zone-rule "PHT-8")

#+end_src

** Locale
UTF-8
#+begin_src emacs-lisp

(set-language-environment "UTF-8")
(set-default-coding-systems 'utf-8)
(set-keyboard-coding-system 'utf-8-unix)

#+end_src

** Bookmarks
Bookmark directory
#+begin_src emacs-lisp

(setq bookmark-default-file "~/.emacs.d/bookmarks")

#+end_src

** Colors
Rainbow mode
#+begin_src emacs-lisp

(use-package rainbow-mode
	:ensure t)

(rainbow-mode 1)

#+end_src

** Emacsclient
Patch: Emacsclient not spawning on dashboard
#+begin_src emacs-lisp

(setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))

#+end_src

* DIY Todo 
- Auto src code block completion (elisp)
- Bind to increase default face attribute (window size)



modules/home-manager/emacs/config.org
#+TITLE: Ohne's GNU/Emacs Configuration
#+AUTHOR: Ohne 
#+DESCRIPTION: Plain GNU/Emacs Configuration

* Package Management
MELPA (Package Archive) and default package.el management
#+begin_src emacs-lisp

(require 'package)
			
(add-to-list 'package-archives
	'("MELPA" . "https://melpa.org/packages/"))

(package-initialize)

(unless (package-installed-p 'use-package)
	(package-refresh-contents)
	(package-install 'use-package))

(use-package magit
	:ensure t)

#+end_src

* Appearance
TODO: Modeline, Scaling
** Themes
Doom themes
#+begin_src emacs-lisp

(use-package doom-themes
	:if window-system
	:ensure t
	:config
	(doom-themes-org-config)
	(doom-themes-visual-bell-config))

(load-theme 'doom-nord t)

#+end_src

** Font
Intel One Mono
#+begin_src emacs-lisp

(when (member "JetBrains Mono" (font-family-list))
	(set-frame-font "JetBrains Mono-18" t t))

(setq font-lock-maximum-decoration t)

#+end_src 

** UI
Cleanup
#+begin_src emacs-lisp

(setq use-dialog-box nil)
(setq use-file-dialog nil)
(setq make-backup-files nil)
(setq auto-save-default nil)
(setq ivy-use-selectable-prompt t)
(menu-bar-mode -1)
(tool-bar-mode -1)
(fringe-mode -1)
(scroll-bar-mode -1)
(global-subword-mode 1)

#+end_src

** Icons
All The Icons
#+begin_src emacs-lisp

(use-package all-the-icons
	:ensure t
	:if (display-graphic-p))

#+end_src

** Line Numbers
Relative line numbers
#+begin_src emacs-lisp

;; (when (version<= "26.0.50" emacs-version )
	;; (global-display-line-numbers-mode)
	;; (setq display-line-numbers-type 'relative))

#+end_src

** Modeline
Doom-modeline.el 
 #+begin_src emacs-lisp

(use-package doom-modeline
	:init
	:ensure t
	:config
	(doom-modeline-mode 1))

(setq doom-modeline-bar-height 5)
(setq doom-modeline-bar-width 1)
(setq doom-modeline-buffer-file-name-style 'truncate-upto-root)
(setq doom-modeline-icon nil)
;; (setq doom-modeline-major-mode-icon t)
;; (setq doom-modeline-major-mode-color-icon t)
;; (setq doom-modeline-minor-modes nil)
(setq display-time-mode t)
(setq doom-modeline-time-icon t)
(setq doom-modeline-time t)
(setq doom-doom-modeline-buffer-name t)
(setq doom-modeline-lsp t)
(setq doom-modeline-buffer-encoding t)
(display-time)

#+end_src

** Cursor
Hl-column/line.el
#+begin_src emacs-lisp


(use-package hl-line
	:ensure t
	:config
	(global-hl-line-mode))

(use-package hl-column
	:ensure t
	:config)

#+end_src
Patch: Org-mode and dashboard breaks hl-column
#+begin_src emacs-lisp

(add-hook 'org-mode-hook (lambda () (hl-column-mode 0)))
(add-hook 'dashboard-mode-hook (lambda () (hl-column-mode 0)))

#+end_src 

** Dashboard
Dashboard.el
#+begin_src emacs-lisp

(use-package page-break-lines
  :ensure t)

(use-package dashboard
  :ensure t
  :config
  (dashboard-setup-startup-hook)
  (setq dashboard-banner-logo-title "in vino veritas")
#  (setq dashboard-startup-banner "~/.config/emacs/images/emacs.png")
  (setq dashboard-page-separator "\n\f\n"))

(global-set-key (kbd "M-o") 'dashboard-open)

#+end_src

** Size
Size and zoomzoom
#+begin_src emacs-lisp

(set-face-attribute 'default nil :height 170)

;; (defun increase-emacs-size () 
;;	(set-face-attribte 'default nil :height (+ x 20)))

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)

#+end_src

* QOL
Quality of Life
** File Completion
Ivy/Counsel.el
#+begin_src emacs-lisp

(electric-pair-mode 1)

(use-package ivy
	:init
	(ivy-mode)
	:ensure t
	:config
	(setq ivy-use-virtual-buffers t)
	(setq enable-recursive-minibuffers t))

(use-package counsel
	:init
	:ensure t)

#+end_src

** Indents/Tabs
Patch: tabs not working ootb
#+begin_src emacs-lisp

(delete-selection-mode 1)
(setq tab-width 4)
(setq-default tab-width 4)
(setq-default indent-tabs-mode nil)

(defun my-insert-tab-char ()
	(interactive)
		(insert "\t"))

(global-set-key (kbd "TAB") 'my-insert-tab-char)

#+end_src

** Binds
Patch: Enable C-u scrolling with Evil
#+begin_src emacs-lisp

(setq evil-want-C-u-scroll t)

#+end_src 
Evil.el 
#+begin_src emacs-lisp

(use-package evil
	:init
	(evil-mode)
	:ensure t
	:config)

#+end_src 
Custom Functions
#+begin_src emacs-lisp

(defun oh/insert-line-below ()
	(interactive)
	(save-excursion
	(end-of-line)
	(open-line 1)))
;
(defun oh/insert-line-above ()
	(interactive)
	(save-excursion
	(end-of-line 0)
	(open-line 1)))

(defun oh/insert-elisp-source-block ()
	(insert "#+begin_src emacs-lisp\n\n#+end_src"))

    #+end_src
General Global Binds
#+begin_src emacs-lisp

(global-set-key (kbd "C-;") 'comment-region)

#+end_src
General.el
#+begin_src emacs-lisp

(use-package general
	:ensure t
	:config
	(general-evil-setup)

	(general-create-definer oh/leader-keys
	:states '(normal insert visual emacs)
	:keymaps 'override
	:prefix "SPC" 
	:global-prefix "M-SPC")
	
  	(oh/leader-keys
    "." '(find-file :wk "Find File")
    "f c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Config")
    "f r" '(counsel-recentf :wk "Find recent files")
	"f s" '(save-buffer :wk "Save file"))

  	(oh/leader-keys
    "b" '(:ignore t :wk "Buffer")
    "b b" '(switch-to-buffer :wk "Switch buffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer")
	"b o" '(evil-buffer-new :wk "New buffer"))
	
	(oh/leader-keys
	"b m" '(bookmark-set :wk "Set bookmark")
	"b M" '(bookmark-delete :wk "Delete bookmark")
	"RET" '(bookmark-jump :wk "Jump to bookmark"))

 	(oh/leader-keys
    "h" '(:ignore t :wk "Help")
    "h f" '(describe-function :wk "Describe function")
    "h v" '(describe-variable :wk "Describe variable")
    "h r r" '(reload-init-file :wk "Reload emacs config")
	"h t" '(load-theme :wk "Load doom theme"))

	(oh/leader-keys
    "t" '(:ignore t :wk "Line Manipulation")
    "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
    "t t" '(visual-line-mode :wk "Toggle truncated lines")
    "t r" '((lambda () (setq display-line-numbers 'relative)) :wk "Toggle line numbers"))

	(oh/leader-keys
	"l" '(:ignore t :wk "Elisp")
	"l e" '(pp-eval-last-sexp :wk "Eval last sexp"))

)

#+end_src
Bind/Key Completion
#+begin_src emacs-lisp

(use-package which-key
	:init
	:ensure t
	:config
	(which-key-mode))

#+end_src 

* Development

** Org (org-mode)
Org.el
#+begin_src emacs-lisp

(use-package org
    :init
    :ensure t
    :config
    (add-hook 'org-mode-hook 'org-indent-mode))

(use-package org-modern
    :after org
    :ensure t
    :config
    (add-hook 'org-mode-hook #'org-modern-mode))

(use-package org-roam
  :ensure t
  :bind (("C-c n l" . org-roam-buffer-toggle)
         ("C-c n f" . org-roam-node-find)
         ("C-c n g" . org-roam-graph)
         ("C-c n i" . org-roam-node-insert)
         ("C-c n c" . org-roam-capture))
  :custom
  (org-roam-capture-templates
	'(
	  ("g" "generic" plain
	  "#+TITLE:  %?"
	  :if-new (file "~/Git/notes/${slug}-%<%m-%d-%Y>.org")
	  :unnarowed t)
         )
  )
  :config
  (org-roam-db-autosync-mode)
  (require 'org-roam-protocol))

(setq org-roam-directory (file-truename "~/Git/notes")) ;; errors when put inside use-package :custom module

;; (use-package org-roam-ui
    ;; :ensure t)

#+end_src
Patch: electric-indent-mode and org-edit-src-content-indentation breaking babel indentation
#+begin_src emacs-lisp

(add-hook 'org-mode-hook 
	(lambda () (electric-indent-local-mode -1)))

(setq org-edit-src-content-indentation 0)

#+end_src

** LSP
lsp-mode
#+begin_src emacs-lisp

(use-package lsp-mode
 	:ensure t)

(use-package lsp-ui
 	:ensure t
 	:commands lsp-ui-mode
 	:hook
 	(lsp-mode . lsp-ui-mode))

(defun org-babel-edit-prep:c (babel-info)
  (setq-local buffer-file-name (->> babel-info caddr (alist-get :tangle)))
  (lsp))

#+end_src
C/C++
#+begin_src emacs-lisp

(use-package ccls
  :ensure t
  :config
  (setq ccls-executable "ccls")
  (setq lsp-prefer-flymake nil)
  (setq-default flycheck-disabled-checkers '(c/c++-clang c/c++-cppcheck c/c++-gcc))
  :hook ((c-mode c++-mode objc-mode) .
         (lambda () (require 'ccls) (lsp))))

#+end_src
Python
#+begin_src emacs-lisp

(use-package lsp-pyright
	:ensure t
	:hook (python-mode . (lambda () (require 'lsp-pyright) (lsp))))

#+end_src

** Completion
Company.el
#+begin_src emacs-lisp

(use-package company
 	:ensure t)

;; (use-package company-lsp
;; 	:ensure t
;; 	:commands company-lsp
;; 	:config
;; 	(push 'company-lsp company-backends))

(use-package flycheck
	:ensure t)

(use-package capf-autosuggest
	:ensure t)

#+end_src

** Project Management
Projectile.el
#+begin_src emacs-lisp

;; Remove parenthesis when done
(use-package projectile
	:ensure t)
	;; :config
	;; (setq projectile-project-search-path '("~/Projects/" )))

#+end_src

** Sudo
Sudo-edit.el
#+begin_src emacs-lisp

;; (use-package sudo-edit
;; 	:config
;; 	(oh/leader-keys
;;   	"fu" '(sudo-edit-find-file :wk "Sudo find file")
;;   	"fU" '(sudo-edit :wk "Sudo edit file")))

#+end_src 

* Misc
Miscellaneous
** Time
Timezone
#+begin_src emacs-lisp

(set-time-zone-rule "PHT-8")

#+end_src

** Locale
UTF-8
#+begin_src emacs-lisp

(set-language-environment "UTF-8")
(set-default-coding-systems 'utf-8)
(set-keyboard-coding-system 'utf-8-unix)

#+end_src

** Bookmarks
Bookmark directory
#+begin_src emacs-lisp

;; (setq bookmark-default-file "~/.config/emacs/bookmarks")

#+end_src

** Colors
Rainbow mode
#+begin_src emacs-lisp

(use-package rainbow-mode
	:ensure t)

(rainbow-mode 1)

#+end_src

** Emacsclient
Patch: Emacsclient not spawning on dashboard
#+begin_src emacs-lisp

(setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))

#+end_src

* DIY Todo 
- Auto src code block completion (elisp)
- Bind to increase default face attribute (window size)



home-manager/emacs/config.el
(require 'package)
			
(add-to-list 'package-archives
	'("MELPA" . "https://melpa.org/packages/"))

(package-initialize)

(unless (package-installed-p 'use-package)
	(package-refresh-contents)
	(package-install 'use-package))

(use-package magit
	:ensure t)

(use-package doom-themes
	:if window-system
	:ensure t
	:config
	(doom-themes-org-config)
	(doom-themes-visual-bell-config))

(load-theme 'doom-nord t)

(when (member "JetBrains Mono" (font-family-list))
	(set-frame-font "JetBrains Mono-18" t t))

(setq font-lock-maximum-decoration t)

(setq use-dialog-box nil)
(setq use-file-dialog nil)
(setq make-backup-files nil)
(setq auto-save-default nil)
(setq ivy-use-selectable-prompt t)
(menu-bar-mode -1)
(tool-bar-mode -1)
(fringe-mode -1)
(scroll-bar-mode -1)
(global-subword-mode 1)

(use-package all-the-icons
	:ensure t
	:if (display-graphic-p))

;; (when (version<= "26.0.50" emacs-version )
	;; (global-display-line-numbers-mode)
	;; (setq display-line-numbers-type 'relative))

(use-package doom-modeline
	:init
	:ensure t
	:config
	(doom-modeline-mode 1))

(setq doom-modeline-bar-height 5)
(setq doom-modeline-bar-width 1)
(setq doom-modeline-buffer-file-name-style 'truncate-upto-root)
(setq doom-modeline-icon nil)
;; (setq doom-modeline-major-mode-icon t)
;; (setq doom-modeline-major-mode-color-icon t)
;; (setq doom-modeline-minor-modes nil)
(setq display-time-mode t)
(setq doom-modeline-time-icon t)
(setq doom-modeline-time t)
(setq doom-doom-modeline-buffer-name t)
(setq doom-modeline-lsp t)
(setq doom-modeline-buffer-encoding t)
(display-time)

(use-package hl-line
	:ensure t
	:config
	(global-hl-line-mode))

(use-package hl-column
	:ensure t
	:config)

(add-hook 'org-mode-hook (lambda () (hl-column-mode 0)))
(add-hook 'dashboard-mode-hook (lambda () (hl-column-mode 0)))

(use-package page-break-lines
  :ensure t)

(use-package dashboard
  :ensure t
  :config
  (dashboard-setup-startup-hook)
  (setq dashboard-banner-logo-title "in vino veritas")
  ;; (setq dashboard-startup-banner "~/.config/emacs/images/emacs.png")
  (setq dashboard-page-separator "\n\f\n"))

(global-set-key (kbd "M-o") 'dashboard-open)

(set-face-attribute 'default nil :height 170)

;; (defun increase-emacs-size () 
;;	(set-face-attribte 'default nil :height (+ x 20)))

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)

(electric-pair-mode 1)

(use-package ivy
	:init
	(ivy-mode)
	:ensure t
	:config
	(setq ivy-use-virtual-buffers t)
	(setq enable-recursive-minibuffers t))

(use-package counsel
	:init
	:ensure t)

(setq warning-minimum-level :error)
(visual-line-mode t)
(global-visual-line-mode t)

(delete-selection-mode 1)
(setq tab-width 4)
(setq-default tab-width 4)

;; (defun my-insert-tab-char ()
;;  	(interactive)
;;  		(insert "\t"))

;; (global-set-key (kbd "C-q") 'my-insert-tab-char)

(setq evil-want-C-u-scroll t)

(use-package evil
	:init
	(evil-mode)
	:ensure t
	:config)

(evil-set-undo-system 'undo-redo)

(defun oh/insert-line-below ()
	(interactive)
	(save-excursion
	(end-of-line)
	(open-line 1)))
;
(defun oh/insert-line-above ()
	(interactive)
	(save-excursion
	(end-of-line 0)
	(open-line 1)))

(defun oh/insert-elisp-source-block ()
	(insert "#+begin_src emacs-lisp\n\n#+end_src"))

(global-set-key (kbd "C-;") 'comment-region)

(use-package general
	:ensure t
	:config
	(general-evil-setup)

	(general-create-definer oh/leader-keys
	:states '(normal insert visual emacs)
	:keymaps 'override
	:prefix "SPC" 
	:global-prefix "M-SPC")
	
  	(oh/leader-keys
    "." '(find-file :wk "Find File")
    "f c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Config")
    "f r" '(counsel-recentf :wk "Find recent files")
	"f s" '(save-buffer :wk "Save file"))

  	(oh/leader-keys
    "b" '(:ignore t :wk "Buffer")
    "b b" '(switch-to-buffer :wk "Switch buffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer")
	"b o" '(evil-buffer-new :wk "New buffer"))
	
	(oh/leader-keys
	"b m" '(bookmark-set :wk "Set bookmark")
	"b M" '(bookmark-delete :wk "Delete bookmark")
	"RET" '(bookmark-jump :wk "Jump to bookmark"))

 	(oh/leader-keys
    "h" '(:ignore t :wk "Help")
    "h f" '(describe-function :wk "Describe function")
    "h v" '(describe-variable :wk "Describe variable")
    "h r r" '(reload-init-file :wk "Reload emacs config")
	"h t" '(load-theme :wk "Load doom theme"))

	(oh/leader-keys
    "t" '(:ignore t :wk "Line Manipulation")
    "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
    "t t" '(visual-line-mode :wk "Toggle truncated lines")
    "t r" '((lambda () (setq display-line-numbers 'relative)) :wk "Toggle line numbers"))

	(oh/leader-keys
	"l" '(:ignore t :wk "Elisp")
	"l e" '(pp-eval-last-sexp :wk "Eval last sexp"))

)

(use-package which-key
	:init
	:ensure t
	:config
	(which-key-mode))

(use-package org
    :init
    :ensure t
    :config
    (add-hook 'org-mode-hook 'org-indent-mode))

(use-package org-modern
    :after org
    :ensure t
    :config
    (add-hook 'org-mode-hook #'org-modern-mode))

(use-package org-roam
  :ensure t
  :bind (("C-c n l" . org-roam-buffer-toggle)
         ("C-c n f" . org-roam-node-find)
         ("C-c n g" . org-roam-graph)
         ("C-c n i" . org-roam-node-insert)
         ("C-c n c" . org-roam-capture))
  :custom
  (org-roam-capture-templates
	'(
	  ("g" "generic" plain
	  "#+TITLE:  %?"
	  :if-new (file "~/Git/notes/${slug}-%<%m-%d-%Y>.org")
	  :unnarowed t)
         ("j" "journal" plain 
	  (file "~/.dotfiles/home-manager/emacs/templates/journal.org")
	  :if-new (file "~/Git/notes/journal-%<%m-%d-%Y>.org")
	  :unnarowed t)
         )
  :config
  (setq org-roam-db-autosync-mode t)
  (require 'org-roam-protocol)))

(setq org-roam-directory "~/Git/notes") ;; errors when put inside use-package :custom module

;; (use-package org-roam-ui
    ;; :ensure t)

;; (add-hook 'org-mode-hook 
;; 	(lambda () (electric-indent-local-mode -1)))

;; (setq org-edit-src-content-indentation 0)

(use-package lsp-mode
 	:ensure t)

(use-package lsp-ui
 	:ensure t
 	:commands lsp-ui-mode
 	:hook
 	(lsp-mode . lsp-ui-mode))

(defun org-babel-edit-prep:c (babel-info)
  (setq-local buffer-file-name (->> babel-info caddr (alist-get :tangle)))
  (lsp))

(use-package ccls
  :ensure t
  :config
  (setq ccls-executable "ccls")
  (setq lsp-prefer-flymake nil)
  (setq-default flycheck-disabled-checkers '(c/c++-clang c/c++-cppcheck c/c++-gcc))
  :hook ((c-mode c++-mode objc-mode) .
         (lambda () (require 'ccls) (lsp))))

(use-package lsp-pyright
	:ensure t
	:hook (python-mode . (lambda () (require 'lsp-pyright) (lsp))))

(use-package company
 	:ensure t)

;; (use-package company-lsp
;; 	:ensure t
;; 	:commands company-lsp
;; 	:config
;; 	(push 'company-lsp company-backends))

(use-package flycheck
	:ensure t)

(use-package capf-autosuggest
	:ensure t)

;; Remove parenthesis when done
(use-package projectile
	:ensure t)
	;; :config
	;; (setq projectile-project-search-path '("~/Projects/" )))

;; (use-package sudo-edit
;; 	:config
;; 	(oh/leader-keys
;;   	"fu" '(sudo-edit-find-file :wk "Sudo find file")
;;   	"fU" '(sudo-edit :wk "Sudo edit file")))

(set-time-zone-rule "PHT-8")

(set-language-environment "UTF-8")
(set-default-coding-systems 'utf-8)
(set-keyboard-coding-system 'utf-8-unix)

(setq bookmark-default-file "~/.emacs.d/bookmarks")

(use-package rainbow-mode
	:ensure t)

(rainbow-mode 1)

(setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))
